File: packages/agent/agent/lib/reminders.ts
import { randomUUID } from "node:crypto";

import { getDb } from "@clara-financas/db";
import { commitments } from "@clara-financas/db/schema/commitment";
import { notifications } from "@clara-financas/db/schema/notification";
import { tenants } from "@clara-financas/db/schema/tenant";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { formatCents } from "@clara-financas/ledger";
import { and, eq } from "drizzle-orm";

import { daysUntil, nextOccurrence, todayInSaoPaulo } from "./dates";

/**
 * O varredor de vencimentos.
 *
 * Vive em `lib/` e não dentro do schedule de propósito: assim pode ser testado
 * e disparado manualmente sem depender do cron, que em `eve dev` nunca roda.
 *
 * **Modo pool:** uma instância atende todos os tenants, então a varredura
 * itera tenants EXPLICITAMENTE e abre um escopo por vez. Nunca há uma consulta
 * "global" — isso furaria o isolamento justamente no caminho automatizado, que
 * é o menos observado. No modo silo (Etapa 4) esta iteração desaparece: cada
 * instância só enxerga o próprio tenant.
 */

export type ReminderResult = {
  today: string;
  tenantsScanned: number;
  created: Array<{ tenantId: string; title: string; dueDate: string; daysUntil: number }>;
};

export async function sweepDueDates(now: Date = new Date()): Promise<ReminderResult> {
  const db = getDb();
  const today = todayInSaoPaulo(now);

  // O registry de tenants fica no control plane e não tem RLS — é lido fora de
  // qualquer escopo de tenant, o que é correto: é o mapa, não o conteúdo.
  const allTenants = await db
    .select({ id: tenants.id })
    .from(tenants)
    .where(eq(tenants.status, "ready"));

  const created: ReminderResult["created"] = [];

  for (const tenant of allTenants) {
    await forTenant(
      tenant.id,
      async (tx) => {
        const rows = await tx
          .select()
          .from(commitments)
          .where(and(eq(commitments.tenantId, tenant.id), eq(commitments.active, "yes")));

        for (const row of rows) {
          const remaining = daysUntil(today, row.dueDate);

          // Fora da janela de aviso ainda; nada a fazer.
          if (remaining > row.remindDaysBefore) continue;

          // Já avisamos deste vencimento. A entrega do cron é at-least-once,
          // e sem esta marca a pessoa receberia o mesmo aviso a cada execução.
          if (row.lastNotifiedFor === row.dueDate) {
            // Vencimento passou e é recorrente: avança para o próximo ciclo.
            if (remaining < 0 && row.recurrenceDayOfMonth !== null) {
              await tx
                .update(commitments)
                .set({ dueDate: nextOccurrence(today, row.recurrenceDayOfMonth) })
                .where(eq(commitments.id, row.id));
            }
            continue;
          }

          const valor =
            row.expectedAmount === null ? "" : ` de ${formatCents(row.expectedAmount)}`;
          const quando =
            remaining < 0
              ? `venceu há ${Math.abs(remaining)} ${Math.abs(remaining) === 1 ? "dia" : "dias"}`
              : remaining === 0
                ? "vence hoje"
                : `vence em ${remaining} ${remaining === 1 ? "dia" : "dias"}`;

          await tx.insert(notifications).values({
            id: `ntf_${randomUUID().replace(/-/g, "").slice(0, 20)}`,
            tenantId: tenant.id,
            kind: "due_date",
            title: `${row.title} ${quando}.`,
            body: `${row.title}${valor} ${quando} (${row.dueDate}). Quer revisar antes de pagar?`,
            commitmentId: row.id,
            createdBy: "process:due_dates",
          });

          await tx
            .update(commitments)
            .set({ lastNotifiedFor: row.dueDate })
            .where(eq(commitments.id, row.id));

          created.push({
            tenantId: tenant.id,
            title: row.title,
            dueDate: row.dueDate,
            daysUntil: remaining,
          });
        }
      },
      db,
    );
  }

  return { today, tenantsScanned: allTenants.length, created };
}

File: services/agent/clara/agents/reminders.py
"""A varredura diária de vencimentos — porta de `agent/lib/reminders.ts` +
`agent/schedules/due_dates.ts`.

Modo pool: uma instância atende todos os tenants, então a varredura itera
tenants EXPLICITAMENTE e abre um escopo por vez — nunca há uma consulta
"global", que furaria o isolamento justamente no caminho automatizado, que é
o menos observado. `tenants` não tem RLS (é o registro do control plane, não
conteúdo de tenant), então é lido fora de qualquer escopo `for_tenant`.

Não usa markdown/task mode: a varredura é determinística, e o modelo só
entra depois, quando a pessoa abre a conversa e pergunta pelos avisos
(`list_notifications`) ou pelos vencimentos (`list_commitments`). Decidir a
quem avisar não é o tipo de coisa que deveria depender de interpretação.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from sqlalchemy import select

from clara.db.models import Commitment, Notification, Tenant
from clara.db.session import get_sessionmaker
from clara.db.tenant_scope import for_tenant
from clara.instructions.dates import days_until, next_occurrence, today_in_sao_paulo
from clara.knowledge.proactivity import proactivity_enabled
from clara.ledger.money import format_cents


@dataclass(frozen=True)
class CreatedReminder:
    tenant_id: str
    title: str
    due_date: str
    days_until: int


@dataclass(frozen=True)
class SweepResult:
    today: str
    tenants_scanned: int
    created: list[CreatedReminder] = field(default_factory=list)


def _ready_tenant_ids() -> list[str]:
    session = get_sessionmaker()()
    try:
        return list(
            session.execute(select(Tenant.id).where(Tenant.status == "ready")).scalars().all()
        )
    finally:
        session.close()


def sweep_due_dates(today: str | None = None) -> SweepResult:
    today_value = today or today_in_sao_paulo()
    tenant_ids = _ready_tenant_ids()
    created: list[CreatedReminder] = []

    for tenant_id in tenant_ids:
        with for_tenant(tenant_id) as session:
            # A metade consentimento: quem desligou os avisos automáticos
            # (`set_proactivity`) não recebe nada, por mais relevante que o
            # vencimento seja.
            if not proactivity_enabled(session, tenant_id):
                continue

            rows = (
                session.execute(
                    select(Commitment).where(
                        Commitment.tenant_id == tenant_id, Commitment.active == "yes"
                    )
                )
                .scalars()
                .all()
            )

            for row in rows:
                remaining = days_until(today_value, row.due_date)

                # Fora da janela de aviso ainda; nada a fazer.
                if remaining > row.remind_days_before:
                    continue

                # Já avisamos deste vencimento. A entrega do scheduler é
                # at-least-once, e sem esta marca a pessoa receberia o mesmo
                # aviso a cada execução.
                if row.last_notified_for == row.due_date:
                    # Vencimento passou e é recorrente: avança para o próximo
                    # ciclo.
                    if remaining < 0 and row.recurrence_day_of_month is not None:
                        row.due_date = next_occurrence(today_value, row.recurrence_day_of_month)
                    continue

                valor = (
                    ""
                    if row.expected_amount is None
                    else f" de {format_cents(row.expected_amount)}"
                )
                if remaining < 0:
                    dias = abs(remaining)
                    quando = f"venceu há {dias} {'dia' if dias == 1 else 'dias'}"
                elif remaining == 0:
                    quando = "vence hoje"
                else:
                    quando = f"vence em {remaining} {'dia' if remaining == 1 else 'dias'}"

                session.add(
                    Notification(
                        tenant_id=tenant_id,
                        kind="due_date",
                        title=f"{row.title} {quando}.",
                        body=(
                            f"{row.title}{valor} {quando} ({row.due_date}). "
                            "Quer revisar antes de pagar?"
                        ),
                        commitment_id=row.id,
                        created_by="process:due_dates",
                    )
                )
                row.last_notified_for = row.due_date

                created.append(
                    CreatedReminder(
                        tenant_id=tenant_id,
                        title=row.title,
                        due_date=row.due_date,
                        days_until=remaining,
                    )
                )

    return SweepResult(today=today_value, tenants_scanned=len(tenant_ids), created=created)
